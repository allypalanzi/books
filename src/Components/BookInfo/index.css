.c-book-info {
  margin: 2%;
}

.c-book-info__container {
  margin: 2% 0;
}

@media(min-width: 600px) {
  .c-book-info__container {
    display: flex;
    text-align: left;
  }
  .c-book-info__image {
    text-align: center;
  }
}

@media(max-width: 600px) {
  .c-book-info__image {
    text-align: center;
    margin: 4% 0;
  }
}

.c-book-info__info {
  padding-left: 4%;
  max-width: 800px;
}

.c-book-info__title {
  font-weight: normal;
  margin-top: 0;
}

.c-book-info__link span, .c-book-info__link svg {
  vertical-align: middle;
}
