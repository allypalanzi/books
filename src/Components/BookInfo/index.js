import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Loader from '../Loader';
import { getBook } from '../../services';
import './index.css';

/*
* A BookInfo component renders information about a book in two ways: either by taking the ID from the url and calling
* the google api, or by being passed props from the component that links to the BookInfo component.
*/

class BookInfo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
    };
  }
  
  async getSingleBook(id) {
    try {
      let data = await getBook(id);
      this.setState({ book: data, loading: false });
    } catch(err) {
      console.error('oh no, an error!', err)
      this.setState({ loading: false });
    }
  }
  
  componentWillMount() {
    if (this.props.location.state) {
      this.setState({ book: this.props.location.state.book, loading: false });
      return;
    } else {
      this.getSingleBook(this.props.match.params.id);
    }    
  }
  
  render() {
    const loading = this.state.loading;
    const book = this.state.book;
    if (loading) {
      return <Loader />;
    } else if (book) {
      return (
        <section className="c-book-info">
          <Link className="c-book-info__link" to="/"><svg className="p-svg" viewBox="0 0 448 512"><path d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"></path></svg> <span className="p-link">Back</span></Link>
          <div className="c-book-info__container">
            <div className="c-book-info__image">
              <img src={book.volumeInfo.imageLinks.thumbnail} alt={`${book.volumeInfo.title} cover`}/>
            </div>
            <div className="c-book-info__info">
              <h2 className="c-book-info__title">{book.volumeInfo.title}</h2>
              <p>{book.volumeInfo.authors.join(', ')}</p>
              <p>{book.volumeInfo.description}</p>
              <a className="p-button" href={book.volumeInfo.previewLink}>Preview</a>
            </div>
          </div>
        </section>
      );
    } else {
      return <p>This book could not be found.</p>
    }
  }
}

export default BookInfo;
