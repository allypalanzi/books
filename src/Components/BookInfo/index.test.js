import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import BookInfo from './';
// MemoryRouter is needed for Link to work
import { MemoryRouter } from 'react-router-dom';

jest.mock('../../services');

describe('<BookInfo/ >', () => {
  const props = {
    location: {},
    match: {
      params: {
        id: 'NaB7oAkon9MC'
      }
    }
  }
  const component = (
    <MemoryRouter>
      <BookInfo {...props} />
    </MemoryRouter>
  );
  const renderedComponent = renderer.create(component);

  it('renders correctly', () => {
    expect(renderedComponent).toMatchSnapshot();
  });
  
  it('shows an error message when the book is not returned', () => {
    const component = shallow(<BookInfo {...props} />);
    component.setState({ book: undefined, loading: false });
    expect(component.find('.test-error').length).toEqual(1);
  });
  
  it('shows Loader when state is loading', () => {
    const component = shallow(<BookInfo {...props} />);
    component.setState({ book: undefined, loading: true });
    expect(component.find('Loader').length).toEqual(1);
  });
});
