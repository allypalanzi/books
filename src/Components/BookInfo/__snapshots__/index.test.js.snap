// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookInfo/ > renders correctly 1`] = `
<section
  className="c-book-info"
>
  <a
    className="c-book-info__link"
    href="/"
    onClick={[Function]}
  >
    <svg
      className="p-svg"
      viewBox="0 0 448 512"
    >
      <path
        d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
      />
    </svg>
     
    <span
      className="p-link"
    >
      Back
    </span>
  </a>
  <div
    className="c-book-info__container"
  >
    <div
      className="c-book-info__image"
    >
      <img
        alt="Statistical Physics cover"
        src="http://books.google.com/books/content?id=NaB7oAkon9MC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      />
    </div>
    <div
      className="c-book-info__info"
    >
      <h2
        className="c-book-info__title"
      >
        Statistical Physics
      </h2>
      <p>
        Lev Davidovich Landau, Evgeniĭ Mikhaĭlovich Lifshit︠s︡, Lev Petrovich Pitaevskiĭ
      </p>
      <p>
        Part 2 of Statistical physics begins with an extensive discussion of the theory of quantum liquids, which was dealt with briefly in the second edition of Statistical physics, by Lev Landau and E.M. Lifshitz; part 1 of Statistical physics is now the third edition of volume 5 of the Course of theoretical physics, by L.D. Landau and E.M. Lifshitz.
      </p>
      <a
        className="p-button"
        href="http://books.google.com/books?id=NaB7oAkon9MC&printsec=frontcover&dq=id:NaB7oAkon9MC%7D&hl=&cd=1&source=gbs_api"
      >
        Preview
      </a>
    </div>
  </div>
</section>
`;
